File: primes.rb
require 'benchmark'

class Integer
  def even?
    self % 2 == 0
  end

  def prime?
    return false if self < 3 or self.even?
    terms = (3..Math.sqrt(self)).reject(&:even?)
    !terms.any? do |term|
      self % term == 0
    end
  end
end

if __FILE__ == $0
  Benchmark.bm(10) do |reporter|
    (4..6).each do |power|
      bound = 10**power
      reporter.report("%9d:" % bound) { (1..bound).each { |x| x.prime? } }
    end
  end
end

File: primes.clj
(use '(clojure.contrib math profile))

(defn odds-below [x]
  (let 
    [lower-bound 3
     upper-bound (+ (floor x) 1)
     step 2]
    (range lower-bound upper-bound step)))
  

(defn prime? [x]
  (if (= x 2)
    true
    (if (or (< x 3) (even? x))
      false
      (every?
        #(not (= (mod x %1) 0))
        (odds-below (sqrt x))))))

(defn primes-below [x]
  (prof :primes
    (filter
      #(prime? %1)
      (range 1 x))))
  
 
;(println (primes-below 100))   
(profile (dotimes [i 5] (last (primes-below 100000))))